examples/ex18.hpp

```diff
@@ -100,55 +108,119 @@ DGHyperbolicConservationLaws::DGHyperbolicConservationLaws(
       nonlinearForm.reset(new NonlinearForm(&vfes));
    }
 #endif
-   nonlinearForm->AddDomainIntegrator(formIntegrator.get());
+   if (preassembleWeakDivergence)
+   {
+      ComputeWeakDivergence();
+   }
+   else
+   {
+      nonlinearForm->AddDomainIntegrator(formIntegrator.get());
+   }
    nonlinearForm->AddInteriorFaceIntegrator(formIntegrator.get());
    nonlinearForm->UseExternalIntegrators();
 
 }
 
 void DGHyperbolicConservationLaws::ComputeInvMass()
 {
-   DenseMatrix Me;     // auxiliary local mass matrix
-   MassIntegrator mi;  // mass integrator
-   // resize it to the current number of elements
-   Me_inv.resize(vfes.GetNE());
-   for (int i = 0; i < vfes.GetNE(); i++)
+   InverseIntegrator inv_mass(new MassIntegrator());
+
+   invmass.resize(vfes.GetNE());
+   for (int i=0; i<vfes.GetNE(); i++)
+   {
+      int dof = vfes.GetFE(i)->GetDof();
+      invmass[i].SetSize(dof);
+      inv_mass.AssembleElementMatrix(*vfes.GetFE(i),
+                                     *vfes.GetElementTransformation(i), invmass[i]);
+   }
+}
+
+void DGHyperbolicConservationLaws::ComputeWeakDivergence()
+{
+   TransposeIntegrator weak_div(new GradientIntegrator());
+   DenseMatrix weakdiv_bynodes;
+
+   weakdiv.resize(vfes.GetNE());
+   for (int i=0; i<vfes.GetNE(); i++)
    {
-      Me.SetSize(vfes.GetFE(i)->GetDof());
-      mi.AssembleElementMatrix(*vfes.GetFE(i),
-                               *vfes.GetElementTransformation(i), Me);
-      DenseMatrixInverse inv(&Me);
-      inv.Factor();
-      inv.GetInverseMatrix(Me_inv[i]);
+      int dof = vfes.GetFE(i)->GetDof();
+      weakdiv_bynodes.SetSize(dof, dof*dim);
+      weak_div.AssembleElementMatrix2(*vfes.GetFE(i), *vfes.GetFE(i),
+                                      *vfes.GetElementTransformation(i), weakdiv_bynodes);
+      weakdiv[i].SetSize(dof, dof*dim);
+      // Reorder so that trial space is ByDim.
+      // This makes applying weak divergence to flux value simpler.
+      for (int j=0; j<dof; j++)
+      {
+         for (int d=0; d<dim; d++)
+         {
+            weakdiv[i].SetCol(j*dim + d, weakdiv_bynodes.GetColumn(d*dof + j));
+         }
+      }
+
    }
 }
 
+
 void DGHyperbolicConservationLaws::Mult(const Vector &x, Vector &y) const
 {
    // 0. Reset wavespeed computation before operator application.
    formIntegrator->ResetMaxCharSpeed();
    // 1. Create the vector z with the face terms (F(u), grad v) - <F.n(u), [w]>.
+   //    If weak divergence is pre-assembled, <F.n(u), [w]>
    nonlinearForm->Mult(x, z);
-   max_char_speed = formIntegrator->GetMaxCharSpeed();
-
-   // 2. Multiply element-wise by the inverse mass matrices.
-   Vector zval;             // local dual vector storage
-   Array<int> vdofs;        // local degrees of freedom storage
-   DenseMatrix zmat, ymat;  // local dual vector storage
-
-   for (int i = 0; i < vfes.GetNE(); i++)
+   if (!weakdiv.empty()) // if weak divergence is pre-assembled
```

Comment:
I think this comment is meant to say "if weak divergence is _not_ pre-assembled"

Reply:
Actually, I think the comment was correct before: "weakdiv is not empty" <-> "weakdiv is precomputed".

Reply:
Oh, my bad. I didn't see that there is `!` when I revise the code. I revert it back the comment and added more comments on auxiliary variables. Please check 5699100631c3e6043263c4cdacc0e5d3b26116fb